reactive/kotlinx-coroutines-jdk9/test/IntegrationTest.kt

```diff
@@ -0,0 +1,150 @@
+        assertThat(pub.awaitFirstOrDefault("OK"), IsEqual("OK"))
+        assertThat(pub.awaitFirstOrNull(), IsNull())
+        assertThat(pub.awaitFirstOrElse { "ELSE" }, IsEqual("ELSE"))
+        assertNSE { pub.awaitLast() }
+        assertNSE { pub.awaitSingle() }
+        var cnt = 0
+        pub.collect { cnt++ }
+        assertThat(cnt, IsEqual(0))
+    }
+
+    @Test
+    fun testSingle() = runBlocking {
+        val pub = publish(ctx(coroutineContext)) {
+            if (delay) delay(1)
+            send("OK")
+        }
+        assertThat(pub.awaitFirst(), IsEqual("OK"))
+        assertThat(pub.awaitFirstOrDefault("!"), IsEqual("OK"))
+        assertThat(pub.awaitFirstOrNull(), IsEqual("OK"))
+        assertThat(pub.awaitFirstOrElse { "ELSE" }, IsEqual("OK"))
+        assertThat(pub.awaitLast(), IsEqual("OK"))
+        assertThat(pub.awaitSingle(), IsEqual("OK"))
+        var cnt = 0
+        pub.collect {
+            assertThat(it, IsEqual("OK"))
+            cnt++
+        }
+        assertThat(cnt, IsEqual(1))
+    }
+
+    @Test
+    fun testNumbers() = runBlocking<Unit> {
+        val n = 100 * stressTestMultiplier
+        val pub = publish(ctx(coroutineContext)) {
+            for (i in 1..n) {
+                send(i)
+                if (delay) delay(1)
+            }
+        }
+        assertThat(pub.awaitFirst(), IsEqual(1))
+        assertThat(pub.awaitFirstOrDefault(0), IsEqual(1))
+        assertThat(pub.awaitLast(), IsEqual(n))
+        assertThat(pub.awaitFirstOrNull(), IsEqual(1))
+        assertThat(pub.awaitFirstOrElse { 0 }, IsEqual(1))
+        assertIAE { pub.awaitSingle() }
+        checkNumbers(n, pub)
+        val flow = pub.asFlow()
+        checkNumbers(n, flow.flowOn(ctx(coroutineContext)).asPublisher())
+    }
+
+    @Test
+    fun testCancelWithoutValue() = runTest {
+        val job = launch(Job(), start = CoroutineStart.UNDISPATCHED) {
+            publish<String> {
+                hang {}
+            }.awaitFirst()
+        }
+
+        job.cancel()
+        job.join()
+    }
+
+    @Test
+    fun testEmptySingle() = runTest(unhandled = listOf({e -> e is NoSuchElementException})) {
+        expect(1)
+        val job = launch(Job(), start = CoroutineStart.UNDISPATCHED) {
+            publish<String> {
+                yield()
+                expect(2)
+                // Nothing to emit
+            }.awaitFirst()
+        }
+
+        job.join()
+        finish(3)
+    }
+
+    private suspend fun checkNumbers(n: Int, pub: Publisher<Int>) {
+        var last = 0
+        pub.collect {
+            assertThat(it, IsEqual(++last))
+        }
+        assertThat(last, IsEqual(n))
+    }
+
+    private inline fun assertIAE(block: () -> Unit) {
+        try {
+            block()
+            expectUnreached()
+        } catch (e: Throwable) {
+            assertThat(e, IsInstanceOf(IllegalArgumentException::class.java))
+        }
+    }
+
+    private inline fun assertNSE(block: () -> Unit) {
```

Comment:
Kotlin has `assrtFailsWith<NoSuchElementException>` instead

Reply:
I didn't actually write any of these tests, just copied what was in `kotlinx-coroutines-reactive` with little change. Should I fix it there, too, in this PR?

Reply:
Yes, it would be great to fix tests there, too.

Reply:
I replaced such methods where I found them.